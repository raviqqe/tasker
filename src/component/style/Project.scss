@import "../../style/colors.scss";
@import "./TextButton.scss";

$name-size: 1.2em;

.Project {
	@include TextButton;

	font-size: $name-size;

	&-disabled {
		@include TextButton($red, $red, transparent);

		font-size: $name-size;
	}
}
