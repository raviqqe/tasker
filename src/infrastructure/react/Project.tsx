import { styled } from "@linaria/react";
import { forwardRef, type Ref } from "react";
import {
  MdArchive,
  MdUnarchive,
  MdDelete,
  MdEdit,
} from "react-icons/md/index.js";
import type * as domain from "../../domain.js";
import { IconButton } from "./IconButton.js";
import { black, red } from "./style/colors.js";

const Container = styled.div`
  display: flex;
  justify-content: space-between
  align-items: center;
  font-size: 1.25em;
`;

const Name = styled.div<{ highlighted: boolean }>`
  word-break: break-word;
  margin-right: 1em;
  flex: 1;
  cursor: ${({ onClick }) => (onClick ? "pointer" : "auto")};
  color: ${({ highlighted }) => (highlighted ? red : black)};
`;

const ButtonsContainer = styled.div`
  display: flex;
  align-items: center;

  > :not(:first-child) {
    margin-left: 0.25rem;
  }
`;

export interface Props {
  archiveProject?: (
    project: domain.Project,
    currentProjectId: string,
  ) => Promise<void>;
  currentProject?: domain.Project;
  deleteProject?: (project: domain.Project) => Promise<void>;
  project: domain.Project;
  switchCurrentProject?: (project: domain.Project) => Promise<void>;
  unarchiveProject?: (project: domain.Project) => Promise<void>;
  updateProject?: (project: domain.Project) => Promise<void>;
}

const ProjectWithRef = (
  {
    archiveProject,
    currentProject,
    deleteProject,
    project,
    switchCurrentProject,
    unarchiveProject,
    updateProject,
  }: Props,
  ref: Ref<HTMLDivElement>,
) => (
  <Container ref={ref}>
    <Name
      highlighted={!!currentProject && project.id === currentProject.id}
      onClick={switchCurrentProject && (() => switchCurrentProject(project))}
    >
      {project.name}
    </Name>
    <ButtonsContainer>
      {updateProject && (
        <IconButton
          aria-label="Update Project"
          onClick={async () => {
            const name = window.prompt("New project name?", project.name);

            if (!name) {
              return;
            }

            await updateProject({ ...project, name });
          }}
        >
          <MdEdit />
        </IconButton>
      )}
      {archiveProject && (
        <IconButton
          aria-label="Archive Project"
          onClick={async () => {
            if (currentProject) {
              await archiveProject(project, currentProject.id);
            }
          }}
        >
          <MdArchive />
        </IconButton>
      )}
      {unarchiveProject && (
        <IconButton
          aria-label="Unarchive Project"
          onClick={() => unarchiveProject(project)}
        >
          <MdUnarchive />
        </IconButton>
      )}
      {deleteProject && (
        <IconButton
          aria-label="Delete Project"
          onClick={() => deleteProject(project)}
        >
          <MdDelete />
        </IconButton>
      )}
    </ButtonsContainer>
  </Container>
);

export const Project = forwardRef(ProjectWithRef);
