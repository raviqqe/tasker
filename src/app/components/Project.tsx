import { styled } from "@linaria/react";
import { forwardRef, type Ref } from "react";
import { useNavigate } from "react-router";
import type * as domain from "../../domain.js";
import { currentProjectSwitcher } from "../../main/current-project-switcher.js";
import { black, red } from "../style.js";

const Container = styled.div`
  display: flex;
  justify-content: space-between
  align-items: center;
  font-size: 1.25em;
`;

const Name = styled.div<{ highlighted: boolean }>`
  word-break: break-word;
  margin-right: 1em;
  flex: 1;
  cursor: ${({ onClick }) => (onClick ? "pointer" : "auto")};
  color: ${({ highlighted }) => (highlighted ? red : black)};
`;

const ButtonsContainer = styled.div`
  display: flex;
  align-items: center;

  > :not(:first-child) {
    margin-left: 0.25rem;
  }
`;

export interface Props {
  buttons?: JSX.Element;
  currentProject?: domain.Project;
  project: domain.Project;
}

export const Project = forwardRef(
  ({ buttons, currentProject, project }: Props, ref: Ref<HTMLDivElement>) => {
    const navigate = useNavigate();

    return (
      <Container ref={ref}>
        <Name
          highlighted={project.id === currentProject?.id}
          onClick={async () => {
            await currentProjectSwitcher.switch(project);
            await navigate("/tasks");
          }}
        >
          {project.name}
        </Name>
        <ButtonsContainer>{buttons}</ButtonsContainer>
      </Container>
    );
  },
);
Project.displayName = "Project";
