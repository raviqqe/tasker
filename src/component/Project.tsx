import * as React from "react";

import NoBoxButton from "./NoBoxButton";
import "./style/Project.css";

interface IProps {
    disabled?: boolean;
    name: string;
    onClick: () => void;
}

export default class Project extends React.Component<IProps> {
    public render() {
        const { disabled, name, onClick } = this.props;

        return (
            <NoBoxButton
                className={"Project-name" + (disabled ? "-disabled" : "")}
                onClick={onClick}
            >
                {name}
            </NoBoxButton>
        );
    }
}
