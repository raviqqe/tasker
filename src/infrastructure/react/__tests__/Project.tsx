import { create } from "react-test-renderer";
import React from "react";
import { Project } from "../Project";

it("renders", () => {
  expect(
    create(<Project project={{ archived: false, id: "", name: "" }} />).toJSON()
  ).toMatchSnapshot();
});
