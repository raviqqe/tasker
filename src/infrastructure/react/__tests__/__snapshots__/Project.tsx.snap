// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="sc-bwzfXH iKmZFG"
>
  <div
    className="sc-htpNat QIOFt"
  >
    
  </div>
  <div
    className="sc-bxivhb dBZEqL"
  />
</div>
`;
